cuqipy_fenics/pde.py

```diff
@@ -106,22 +226,104 @@ def gradient_wrt_parameter(self):
     def _create_observation_operator(self, observation_operator):
         raise NotImplementedError
 
+    def _is_parameter_updated(self, value):
+        """ A helper function to check if the PDE model parameter (the parameter
+         to be inferred) is updated """
+
+        if hasattr(self, '_parameter') \
+            and np.allclose(self._parameter.vector().get_local(),
+                            value.vector().get_local(),
+                            atol=dl.DOLFIN_EPS, rtol=dl.DOLFIN_EPS):
+            return False
+        else:
+            return True
+
+
 class SteadyStateLinearFEniCSPDE(FEniCSPDE):
     """ Class representation of steady state linear PDEs defined in FEniCS. It accepts the same arguments as the base class `cuqipy_fenics.pde.FEniCSPDE`."""
 
     def assemble(self, parameter=None):
+        self._solution_trial_function = dl.TrialFunction(
+            self.solution_function_space)
+        self._solution_test_function = dl.TestFunction(
+            self.solution_function_space)
+
         if parameter is not None:
             self.parameter = parameter
 
-        solution_trial_function = dl.TrialFunction(self.solution_function_space)
-        solution_test_function = dl.TestFunction(self.solution_function_space)
-        self.diff_op, self.rhs  = \
-            dl.lhs(self.PDE_form(self.parameter, solution_trial_function,solution_test_function)),\
-            dl.rhs(self.PDE_form(self.parameter, solution_trial_function, solution_test_function))
+        # Either assemble the lhs and rhs forms separately or the full PDE form
+        if self.lhs_form is not None:
+            self._assemble_lhs()
+            self._assemble_rhs()
+
+        else:
+            self._assemble_full()
+
+
+
+    def with_updated_rhs(self, rhs_form):
+        """ A method to create a shallow copy of the PDE model with updated rhs 
```

Comment:
Perhaps we should add a warning here checking if reuse_assembled is not True. If so, tell the user that they wont reuse factorizations.

Reply:
Great! added!